### PR TITLE
fix the scale dot attention doc

### LINKED ISSUE #120810: Scaled-Dot-Product-Attention Has Broadcastable Mask
### 📚 The doc issue

The [scaled-dot-product-attention doc](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html) claims
```
attn_mask (optional Tensor) – Attention mask; shape (N,...,L,S). 
```
However, if we try the attached [broadcast-mask.py.zip](https://github.com/pytorch/pytorch/files/14438879/broadcast-mask.py.zip), we will see
```
query shape = torch.Size([3, 2, 4, 7, 16])
key shape = torch.Size([3, 2, 4, 5, 16])
value shape = torch.Size([3, 2, 4, 5, 32])
mask shape = torch.Size([1, 1, 5])
attention shape = torch.Size([3, 2, 4, 7, 32])
```
This means our real shape requirement for `attn_mask` is `(..., S)` and broadcastable to `(N, ..., L, S)`

### Suggest a potential alternative/fix

Please update the [scaled-dot-product-attention doc](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html) so it is consistent with the code behavior.

Or, if the code behavior is to-be-changed to conform to doc, please either raise exception in code or mark deprecation warning.

cc @svekars @brycebortree @jbschlosser @bhosmer @cpuhrsch @erichan1 @drisspg @mikaylagawarecki

### PR DESCRIPTION
**Pull request by lancerts**
Fixes #120810

The code verifies the broadcast behavior (from the issue),
```
import torch

B = 3
S = 5
L = 7
E = 16
EV = 32
additional_batches = [2, 4]

query_shape = [B] + additional_batches + [L, E]
key_shape = [B] + additional_batches + [S, E]
value_shape = [B] + additional_batches + [S, EV]


query = torch.rand(*query_shape)
key = torch.rand(*key_shape)
value = torch.rand(*value_shape)
mask = torch.zeros((1, 1, S), dtype=torch.bool)
mask[:, :, S // 2 :] = True

# query.to("cuda")
# key.to("cuda")
# value.to("cuda")
# mask.to("cuda")


attention = torch.nn.functional.scaled_dot_product_attention(query, key, value, mask)

print(f"query shape = {query.shape}")
print(f"key shape = {key.shape}")
print(f"value shape = {value.shape}")
print(f"mask shape = {mask.shape}")
print(f"attention shape = {attention.shape}")

#in both CPU and cuda, output shape is:
# query shape = torch.Size([3, 2, 4, 7, 16])
# key shape = torch.Size([3, 2, 4, 5, 16])
# value shape = torch.Size([3, 2, 4, 5, 32])
# mask shape = torch.Size([1, 1, 5])
# attention shape = torch.Size([3, 2, 4, 7, 32])

## test add is broadcasting mask to query@(key.mT)
res = query@(key.mT)
print(res.shape)
res2 = torch.add(res, mask)
print(res2.shape)
```

At code level, in the default backend,
https://github.com/pytorch/pytorch/blob/ab38354887fe86e611f6f5bef0b9d7cf72e27d8b/aten/src/ATen/native/transformers/attention.cpp#L735

the add operation is broadcasting the `attn_mask` to `auto attn = at::matmul(query, key.transpose(-2, -1) * scaling_factor);`

- Changed the doc in [torch/nn/functional.py](https://github.com/pytorch/pytorch/pull/120859/files#diff-c358c214f663ba0c8b9c6846fbe0042fa29494cf02fe4714a17dcd0d268b035b).
- Also fixed a few inconsistencies in the cpp comments.

@mikaylagawarecki

### CODE REVIEW

**Comment by YifanShenSZ** [comment]
Hi @lancerts, many thanks for the prompt fix! The "`(..., S)` and broadcastable to `(N, ..., L, S)`" shape requirement for `attn_mask` is what I inferred. If the source code is
```
auto attn = at::matmul(query, key.transpose(-2, -1) * scaling_factor);
attn = attn.add(*attn_mask);
```
Do you think it only requires "broadcastable to `(N, ..., L, S)`"?

**Comment by lancerts** [comment]
> Hi @lancerts, many thanks for the prompt fix! The "`(..., S)` and broadcastable to `(N, ..., L, S)`" shape requirement for `attn_mask` is what I inferred. If the source code is
> 
> ```
> auto attn = at::matmul(query, key.transpose(-2, -1) * scaling_factor);
> attn = attn.add(*attn_mask);
> ```
> 
> Then it might only require "broadcastable to `(N, ..., L, S)`"? Unless there is another shape checking somewhere? Wdyt?

AFAIK, it is the case.

Further tested two cases on both CPU and GPU, both passed
```
mask = torch.zeros((1, 1, 1), dtype=torch.bool)
mask[:, :, S // 2 :] = True
```
and

```
mask = torch.zeros((1), dtype=torch.bool)
mask[0] = True
```

**Review by drisspg — APPROVED** [review]
Thanks!

@pytorchbot merge

**Comment by lancerts** [comment]
@drisspg amend the indent to resolve doc build issue.
